protocol/xmrtaker/event.go

```diff
@@ -0,0 +1,229 @@
+package xmrtaker
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/athanorlabs/atomic-swap/common/types"
+	mcrypto "github.com/athanorlabs/atomic-swap/crypto/monero"
+	"github.com/athanorlabs/atomic-swap/net/message"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+// getStatus returns the status corresponding to the next expected event.
+func getStatus(t Event) types.Status {
+	switch t.(type) {
+	case *EventXMRLocked:
+		return types.ETHLocked
+	case *EventETHClaimed:
+		return types.ContractReady
+	default:
+		return types.UnknownStatus
+	}
+}
+
+// Event represents a swap state event.
+type Event interface{}
+
+// EventKeysReceived is the first expected event.
+type EventKeysReceived struct {
+	message *message.SendKeysMessage
+	errCh   chan error
+}
+
+func newEventKeysReceived(msg *message.SendKeysMessage) *EventKeysReceived {
+	return &EventKeysReceived{
+		message: msg,
+		errCh:   make(chan error),
+	}
+}
+
+// EventXMRLocked is the second expected event. It represents XMR being locked
+// on-chain.
+type EventXMRLocked struct {
+	message *message.NotifyXMRLock
+	errCh   chan error
+}
+
+func newEventXMRLocked(msg *message.NotifyXMRLock) *EventXMRLocked {
+	return &EventXMRLocked{
+		message: msg,
+		errCh:   make(chan error),
+	}
+}
+
+// EventETHClaimed is the third expected event. It represents the ETH being claimed
+// tbyo the counterparty, and thus we can also claim the XMR.
+type EventETHClaimed struct {
+	sk    *mcrypto.PrivateSpendKey
+	errCh chan error
+}
+
+func newEventETHClaimed(sk *mcrypto.PrivateSpendKey) *EventETHClaimed {
+	return &EventETHClaimed{
+		sk:    sk,
+		errCh: make(chan error),
+	}
+}
+
+// EventShouldRefund is an optional event. It occurs when the XMR-maker doesn't
+// lock before t0, so we should refund the ETH.
+type EventShouldRefund struct {
+	errCh    chan error
+	txHashCh chan ethcommon.Hash // contains the refund tx hash, if successful
+}
+
+func newEventShouldRefund() *EventShouldRefund {
+	return &EventShouldRefund{
+		errCh:    make(chan error),
+		txHashCh: make(chan ethcommon.Hash, 1),
+	}
+}
+
+// EventExit is an optional event. It is sent when the protocol should be stopped,
+// for example if the remote peer closes their connection with us before sending all
+// required messages, or we decide to cancel the swap.
+type EventExit struct {
+	errCh chan error
+}
+
+func newEventExit() *EventExit {
+	return &EventExit{
+		errCh: make(chan error),
+	}
+}
+
+func (s *swapState) runHandleEvents() {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case event := <-s.eventCh:
+			s.handleEvent(event)
+		}
+	}
+}
+
+func (s *swapState) handleEvent(event Event) {
+	// events are only used once, so their error channel can be closed after handling.
+	switch e := event.(type) {
+	case *EventKeysReceived:
+		log.Infof("EventKeysReceived")
+		defer close(e.errCh)
+
+		if reflect.TypeOf(s.nextExpectedEvent) != reflect.TypeOf(&EventKeysReceived{}) {
```

Comment:
I assume we are using reflection because it is awkward to check if `s.nextExpectedEvent` is nil and `.(type)` will panic on nil?  Maybe in the future we could have some sort of `None` event and not allow nil? Formatted prints, like the one on the next line, would also be improved.

Reply:
yeah I was going to change the `Event` interface to have a `Type()` function, would make these checks nicer. will add